game-core/src/test/java/games/strategy/triplea/delegate/battle/casualty/CasualtyOrderOfLossesTestOnBigWorldV3.java

```diff
@@ -165,4 +161,115 @@ void infantryAndArtillery() {
+                .enemyUnits(List.of())
+                .battlesite(testData.france)
+                .costs(testData.costMap)
+                .data(testData.gameData)
+                .build());
+
+    assertThat(result, hasSize(4));
+    assertThat(
+        "Non amphibious marine only has attack of 2 since it doesn't get marine bonus",
+        result.get(0),
+        is(attackingUnits.get(3)));
+    assertThat(result.get(1), is(attackingUnits.get(1)));
+    assertThat("Amphibious marine has attack of 3", result.get(2), is(attackingUnits.get(2)));
+    assertThat(result.get(3), is(attackingUnits.get(0)));
+  }
+
+  @Test
+  void amphibiousAndNonAmphibiousCaching() {
+    testData.addTech(new ImprovedArtillerySupportAdvance(testData.gameData));
+
+    final List<Unit> amphibUnits = new ArrayList<>();
+    amphibUnits.addAll(testData.tank(1));
+    amphibUnits.addAll(testData.artillery(1));
+    amphibUnits.addAll(testData.marine(1));
+
+    amphibUnits.forEach(
+        unit -> {
+          unit.getProperty(Unit.UNLOADED_AMPHIBIOUS)
+              .ifPresent(
+                  property -> {
+                    try {
+                      property.setValue(true);
+                    } catch (final MutableProperty.InvalidValueException e) {
+                      // should not happen
+                    }
+                  });
+        });
+
+    final List<Unit> attackingUnits = new ArrayList<>(amphibUnits);
+    attackingUnits.addAll(testData.marine(1));
+
+    final List<Unit> result =
+        CasualtyOrderOfLosses.sortUnitsForCasualtiesWithSupport(
+            CasualtyOrderOfLosses.Parameters.builder()
+                .targetsToPickFrom(attackingUnits)
+                .combatModifiers(
+                    CombatModifiers.builder().defending(false).territoryEffects(List.of()).build())
+                .player(testData.british)
+                .enemyUnits(List.of())
+                .battlesite(testData.france)
+                .costs(testData.costMap)
+                .data(testData.gameData)
+                .build());
+
+    assertThat(
+        "Non amphibious marine only has attack of 2 since it doesn't get marine bonus",
+        result.get(0),
+        is(attackingUnits.get(3)));
+    assertThat("Amphibious marine has attack of 3", result.get(2), is(attackingUnits.get(2)));
+
+    final List<Unit> result2 =
+        CasualtyOrderOfLosses.sortUnitsForCasualtiesWithSupport(
+            CasualtyOrderOfLosses.Parameters.builder()
+                .targetsToPickFrom(attackingUnits.subList(0, 3))
+                .combatModifiers(
+                    CombatModifiers.builder().defending(false).territoryEffects(List.of()).build())
+                .player(testData.british)
+                .enemyUnits(List.of())
+                .battlesite(testData.france)
+                .costs(testData.costMap)
+                .data(testData.gameData)
+                .build());
+
+    assertThat(result2, hasSize(3));
```

Comment:
I'd really like to test that the cache was used and that the rest of the code was skipped, but it is all static based so I can't `spy` on method calls to see what is called or not.  I manually stepped through the code for this test to verify that the cache was being used.

Reply:
IMO that is good enough. Though I *think* the cache is actually only used once. All of this code is called to determine where units are to be placed on the combat board, and then later we go through this path with cache when computing supports.

I took a deep look at calculating supports some time ago but the performance was not super easy to match. If you're ever interested I could go over you where I left and what I think is a simple algorithm for this.